### PR TITLE
fuse3: init at 3.1.1

### PR DESCRIPTION
**Pull request by primeos**
###### Motivation for this change

Add `fuse3` (libfuse 3.x). Any help is welcome :smile: - I'm not that familiar with `libfuse`...

> ## Note to Packagers
>
> libfuse 3 is designed to be co-installable with libfuse 2. However, some files will be installed by both libfuse 2 and libfuse 3 (e.g. /etc/fuse.conf, the udev and init scripts, and the
mount.fuse(8) manpage). These files should be taken from libfuse 3. The format/content is guaranteed to remain backwards compatible with libfuse 2.
>
> We recommend to ship libfuse2 and libfuse3 in three separate packages: a libfuse-common package that contains files shared by libfuse 2+3 (taken from the libfuse3 tarball), and libfuse2 and
libfuse3 packages that contain the shared library and helper programs for the respective version.

### TODO:
- [x] Install libfuse 2 and 3 by default
- [x] /etc/fuse.conf from libfuse 3
  - We currently don't provide /etc/fuse.conf
- [x] udev and init scripts from libfuse 3
  - [x] udev rules from libfuse 3
  - [x] Init scripts not included
- [x] `mount.fuse` man page from libfuse 3 (~~TODO: define priority~~)
- [x] Add `libfuse-common` like recommended (added but not really used as own package)?
- [x] Merge the package definitions for 2.x and 3.x (e.g. common.nix, 2.nix and 3.nix)?
- [ ] Wait for some reviews (some things might be wrong/missing)

@mornfall would you like to maintain `libfuse3` as well?

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
Do you know a file system already using fuse3?

**Comment by primeos** [comment]
@Mic92 Yes, I've tested it with `sshfs-fuse` (version 3.2.0). But I just realized that it might be the only one so far...

I'll add that commit to this PR.

**Review by Mic92 — APPROVED** [review]
I can use the old sshfs as well as the new version in parallel.

**Comment by primeos** [comment]
@Mic92 Sorry for my late reply & thanks for testing it. If you don't mind I'll resolve the last TODOs (might have to wait until next week) and merge it after the branch off, since it isn't that important ATM and `sshfs 2.10` is probably better for `17.09` anyway.

> I can use the old sshfs as well as the new version in parallel.

I would've replaced sshfs 2.x with 3.x, because of the following note from the 2.10 release:
> This is a maintenance release of the SSHFS 2.x line.
Please use a SSHFS 3.x release if you can!

Is that ok or should I keep sshfs 2.x and add a new attribute for 3.x (`sshfs3`)?

**Comment by fpletz** [comment]
Seems fine to me. Thanks!

**Comment by primeos** [comment]
I've just updated the PR. Now all files provided by libfuse 2 and 3 should come from libfuse 3 as recommended (just in case). I've also added `fuse-common` (might be useful for udev) but I couldn't figure out a good way to use it for `fuse` and `fuse3` (since the latter would obviously cause an infinite recursion or I would have to use another approach with other disadvantages). Imho replacing the shared files from libfuse 2 with the ones from libfuse 3 / fuse-common and using priorities to avoid collision warnings was the best idea, but it isn't perfect... - Any possible suggestions to improve this are welcome :smile:

But at least this way should be backwards compatible. If it looks ok I'd run some final tests and merge it.

**Pull request closed without merging**
